### PR TITLE
Improve autocomplete query behavior

### PR DESCRIPTION
**Pull request by rodrigohenriques**
**What this PR fixes:**
The previous query buffer strategy weren't as efficient as possible. So I created a new operator to filter source emissions only passing the last one to downstream subscribers after a specified time without new emissions. 

**How this PR was tested:**
I've created new unit tests to cover this new operator functionalities and I've updated the existents unit tests according to the new query behavior.

### CODE REVIEW

File: library/src/test/java/com/a99/rxplaces/TestsExtensions.kt [hunk 1]
```diff
@@ -1,8 +1,18 @@
...
...short_name\":\"254\",\"types\":[\"street_number\"]},{\"long_name\":\"Rua Joaquim Floriano\",\"short_name\":\"R. Joaquim Floriano\",\"types\":[\"route\"]},{\"long_name\":\"Itaim Bibi\",\"short_name\":\"Itaim Bibi\",\"types\":[\"political\",\"sublocality\",\"sublocality_level_1\"]},{\"long_name\":\"S\u00E3o Paulo\",\"short_name\":\"S\u00E3o Paulo\",\"types\":[\"administrative_area_level_2\",\"political\"]},{\"long_name\":\"S\u00E3o Paulo\",\"short_name\":\"SP\",\"types\":[\"administrative_area_level_1\",\"political\"]},{\"long_name\":\"Brasil\",\"short_name\":\"BR\",\"types\":[\"country\",\"political\"]},{\"long_name\":\"04534-001\",\"short_name\":\"04534-001\",\"types\":[\"postal_code\"]}],\"formatted_address\":\"R. Joaquim Floriano, 254 - Itaim Bibi, S\u00E3o Paulo - SP, 04534-001, Brasil\",\"geometry\":{\"location\":{\"lat\":-23.5838378,\"lng\":-46.6726088},\"location_type\":\"APPROXIMATE\",\"viewport\":{\"northeast\":{\"lat\":-23.58248881970849,\"lng\":-46.6712598197085},\"southwest\":{\"lat\":-23.5851867802915,\"lng\":-46.6739577802915}}},\"place_id\":\"ChIJWRNvGF9XzpQRetBgux1t6qw\",\"types\":[\"establishment\",\"food\",\"point_of_interest\",\"restaurant\"]}],\"status\":\"OK\"}"
```
Comment (stik3nd) [inline library/src/test/java/com/a99/rxplaces/TestsExtensions.kt]: You can use triple quotes here if you wish.
https://kotlinlang.org/docs/reference/basic-types.html#strings

File: library/src/test/java/com/a99/rxplaces/RxAutocompleteStressTest.kt [hunk 2]
```diff
@@ -43,17 +42,16 @@ class RxAutocompleteStressTest {
 
     // then
     testSubscriber.assertNoErrors()
-    testSubscriber.assertValueCount(3)
+    testSubscriber.assertValueCount(1)
 
-    verify(repository, times(3)).query(any(), any())
+    verify(repository, times(1)).query(any(), any())
```
Comment (rsicarelli) [inline library/src/test/java/com/a99/rxplaces/RxAutocompleteStressTest.kt]: If the `times()` are `1`, you do not need to pass it as parameter. Please change it to `verify(repository).query()`

File: library/src/main/kotlin/com/a99/rxplaces/AutocompleteBufferOperator.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            scheduledCompletion?.invoke()
+          },
+          timespan.interval(), timespan.timeUnit())
+    }
+
+    private fun cancelPreviousEmission() {
+      emitterSubscription.unsubscribe()
+    }
+  }
+}
```
Comment (walmyrcarvalho) [inline library/src/main/kotlin/com/a99/rxplaces/AutocompleteBufferOperator.kt]: I'll create a bot to add an empty line on the end of all of your PR files.

(kidding)

File: library/src/main/kotlin/com/a99/rxplaces/Extensions.kt [hunk 4]
```diff
@@ -9,3 +11,7 @@ internal fun <T> List<T>.toPipedString(): String? {
 internal fun Pair<Double, Double>.formatWithComma(): String {
   return "$first,$second"
 }
+
+internal fun Pair<Long, TimeUnit>.interval() = first
+internal fun Pair<Long, TimeUnit>.timeUnit() = second
+internal fun Pair<Long, TimeUnit>.intervalInMillis() = timeUnit().toMillis(interval())
```
Comment (walmyrcarvalho) [inline library/src/main/kotlin/com/a99/rxplaces/Extensions.kt]: Missing empty line on the end (can create messy merges)

File: library/src/main/kotlin/com/a99/rxplaces/RxAutocomplete.kt [hunk 5]
```diff
@@ -42,9 +42,7 @@ class RxAutocomplete internal constructor(
         .observeOn(scheduler)
         .filter { it.length > minKeyStroke }
         .doOnNext { logger("RxAutocomplete", "Received: $it") }
-        .buffer(queryInterval.first, queryInterval.second)
-        .filter { it.isNotEmpty() }
-        .map { it.last() }
+        .lift(createAutoCompleteBufferOperator())
```
Comment (walmyrcarvalho) [inline library/src/main/kotlin/com/a99/rxplaces/RxAutocomplete.kt]: TIL about this operator. :O

File: build.gradle [hunk 6]
```diff
@@ -1,9 +1,7 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
-  ext.libraryVersion = "0.1.0"
-
-  ext.kotlin_version = '1.1.2-3'
+  ext.kotlin_version = '1.1.2-5'
```
Comment (rodrigohenriques) [inline build.gradle]: We could change it in another PR.

File: library/src/main/kotlin/com/a99/rxplaces/AutocompleteBufferOperator.kt (hunk 3 above)
Comment (rodrigohenriques) [inline library/src/main/kotlin/com/a99/rxplaces/AutocompleteBufferOperator.kt]: I believe that it is a problem from kotlin plugin

File: library/src/main/kotlin/com/a99/rxplaces/RxAutocomplete.kt (hunk 5 above)
Comment (rodrigohenriques) [inline library/src/main/kotlin/com/a99/rxplaces/RxAutocomplete.kt]: Lift is awesome bro! 👍

**Pull request merged**